RAKP Message 1: implement layerexts.LayerDecodingLayer interface

oliverpool opened the pull request:
This PR adds the ability to decode `RAKP Message 1`.

See #44 for some background, why we need this.

---

I set the `NextLayerType` to `gopacket.LayerTypePayload` (like Message 2 and 4), but I have no idea if this is correct :see_no_evil:

gebn commented on pkg/ipmi/rakp_message_1.go:
```diff
@@ -75,6 +75,36 @@ func (r *RAKPMessage1) SerializeTo(b gopacket.SerializeBuffer, opts gopacket.Ser
+	copy(r.RemoteConsoleRandom[:], data[8:24])
+	r.MaxPrivilegeLevel = PrivilegeLevel(data[24] & 0xF)
+	r.PrivilegeLevelLookup = (data[24] & (1 << 4)) == 0
+	lenUsername := data[27]
```
Given this is a numeric length, should we convert to `uint8`?

oliverpool replied:
You are right.

gebn commented on pkg/ipmi/rakp_message_1.go:
```diff
@@ -75,6 +75,36 @@ func (r *RAKPMessage1) SerializeTo(b gopacket.SerializeBuffer, opts gopacket.Ser
+		}
+		if len(data) < int(28+lenUsername) {
+			df.SetTruncated()
+			return fmt.Errorf("RAKP Message 2 not long enough to contain username. Expected %v, got %v", 28+lenUsername, len(data))
```
`Message 2` -> `Message 1`

gebn commented on pkg/ipmi/rakp_message_1.go:
```diff
@@ -75,6 +75,36 @@ func (r *RAKPMessage1) SerializeTo(b gopacket.SerializeBuffer, opts gopacket.Ser
+	r.MaxPrivilegeLevel = PrivilegeLevel(data[24] & 0xF)
+	r.PrivilegeLevelLookup = (data[24] & (1 << 4)) == 0
+	lenUsername := data[27]
+	if lenUsername > 0 {
```
I think the code will still work correctly without this check - let's remove.

oliverpool replied:
indeed!

gebn commented on pkg/ipmi/rakp_message_1_test.go:
```diff
@@ -74,5 +74,30 @@ func TestRAKPMessage1SerializeTo(t *testing.T) {
+		if r.MaxPrivilegeLevel != decoded.MaxPrivilegeLevel {
+			t.Errorf("decode %v MaxPrivilegeLevel = got %v, want %v", test.layer, decoded.MaxPrivilegeLevel, r.MaxPrivilegeLevel)
+		}
+		if r.PrivilegeLevelLookup != decoded.PrivilegeLevelLookup {
```
Nit: can you swap this and the `MaxPrivilegeLevel` comparison to match the order of fields in the `RAKPMessage1` struct.

gebn commented on pkg/ipmi/rakp_message_1_test.go:
```diff
@@ -74,5 +74,30 @@ func TestRAKPMessage1SerializeTo(t *testing.T) {
+			continue
+		}
+		r := test.layer
+		if r.Tag != decoded.Tag {
```
Nit: we want to check whether `decoded.Tag` (the thing that might be incorrect) is equal to `r.Tag` (the reference value), rather than the other way around, so we'd typically swap the operands, i.e. `decoded.Tag != r.Tag`. As written, this is sometimes called a [Yoda condition](https://en.wikipedia.org/wiki/Yoda_conditions).

To ease writing an if-statement for every field, which is easy to get wrong, I originally used `cmp.Diff()` for comparing the entire produced and reference structs - see `rakp_message_2_test.go`.

oliverpool replied:
Thank for the hint to `cmp.Diff`: it makes the test much shorter and any failure much more readable!

I had to add `test.layer.BaseLayer.Contents = got`, but everything seems now correct.

gebn reviewed (changes requested):
Thanks for this! Just a couple of amendments, please see comments.

oliverpool commented:
Thank you for the review!

I addressed your comments and also added "subtests" using `t.Run(test.name, func(t *testing.T) {` (so that a failure also shows which test of the table is concerned).
